### DISCUSSION TITLE
Argon progress bar density may be somewhat difficult to parse (accessibility)

### DISCUSSION THREAD

**Discussion by verarr** [ppy/osu#25625] [Ideas]
## Why?

The density graph in "triangles" is displayed as an area graph, which is very easy to read at a glance, because it's straightforward that a taller graph means more things going on. On the other hand, "argon" only has gray or white in different opacities, which makes it take some time to figure out where there might be difficulty spikes etc. (_For my instance, this may be because of some issues I have with general mental processing, but it still makes some things harder for me and I doubt I'm the only one._)

## Suggestions
The **original screenshot** I will be doing various edits on looks like this:
![cropped screenshot of the "argon" progress bar](https://github.com/ppy/osu/assets/72462464/5287a203-0ca2-4d34-b196-5c1dd26feeba)

### Modify color value curves

The current implementation might be difficult to read because it has low contrast, or in other words, the brightness of the bar is too homogenous.

Editing a screenshot in GIMP, I found that a curve like this helps bring out more important details:

![screenshot of the Levels tool in GIMP](https://github.com/ppy/osu/assets/72462464/8ff74909-ba9d-43d2-aee3-d37ad0d61a3b)

After adjusting levels so that the minimum value is colored black and the maximum value is colored white, then applying the above curve, we get this:

![edited screenshot where the progress bar has way more contrast](https://github.com/ppy/osu/assets/72462464/a13d7885-ec05-46bf-9c1e-4f3c99e3fdc4)

We could obviously make it look better by not using extreme values like pitch black (which blends in with the 100% BG dim), but the actual difficulty spikes now stand out more and are therefore easier to locate.

### Use color instead of opacity

The simplest way to explain this is to have a color in [HSV](https://en.wikipedia.org/wiki/HSL_and_HSV) where the _Hue_ and the _Value_ are constant, and the _Saturation_ ranges from 0% to 100% according to density (sparse → dense). HSL provides an even more flexible solution via the _Lightness_ scale, enabling interpolation between black _or_ white with one value. You can try this in most advanced image editors:

![screenshot of the color picker in GIMP](https://github.com/ppy/osu/assets/72462464/4a33ea66-6236-4112-b4c0-bc0f93a4cc08)

Assuming the _Hue_ is 47° (for a _nice_ orange) and the _Value_ is 100%, this results in a line colored **white** in empty segments/breaks, a **light shade** of orange for segments with medium density and **very vibrant** orange for the most dense parts.

By editing the previous image further, after applying the _Gradient Map_ operation while the FG color is white and the BG color is our color of choice, it might look something like this:

![edited screenshot where the progress bar has a pure white background and orange accents](https://github.com/ppy/osu/assets/72462464/adb6465d-8af2-4423-a551-f648ebdfab69)

The base color of white could be brought down to something less bright, like a **15% gray** (or something that fits the visual design of argon) for a less intrusive bar:

![edited screenshot where the progress bar has a 15% gray background and orange accents](https://github.com/ppy/osu/assets/72462464/9bb2ec2a-9215-405f-a401-14027fa12092)

The _Hue_ should be adjustable (in the skin editor) because some types of colorblindness may make certain colors harder to distinguish from gray (try some colors on [the Coolors Color Picker](https://coolors.co/color-picker)).

The maximum value of the _Saturation_ should also be adjustable, for people who'd prefer if the bar wasn't as visually prevalent.

#### Picking colors dynamically

There's no _one_ color that universally has good contrast with _any_ other arbitrary color. If you'd like to color the bar but also see the background _and_ have good contrast, the bar's color needs to be derived from the background.

For this, we could **sample some colors** from that particular area of the background image where the bar is located. The colors first need to be **inverted**, and if we want to eliminate extreme values, we should apply some sort of blur (in this example I used a Gaussian blur with a radius of 25px, but something simpler would probably totally be ok as well). Here are some examples:

![portion of the background image for "Tower of Heaven" with inverted colors](https://github.com/ppy/osu/assets/72462464/dc28a735-a55a-4cc6-8e03-58bf1be83029)
![portion of the background image for "Airman ga Taosenai" with inverted and oversaturated colors](https://github.com/ppy/osu/assets/72462464/11882d25-fa33-4d00-8db7-918a1e5852f8)
![portion of the background image for "ILY" (by Panda Eyes) with inverted colors](https://github.com/ppy/osu/assets/72462464/57bf1854-bb2a-4005-966c-3a051c62026d)
![portion of the background image for "Ooi" (by O2i3) with inverted and oversaturated colors](https://github.com/ppy/osu/assets/72462464/1c0a5091-4580-4c5d-ae71-db91922f43cf)

Some **adjustments to the saturation and/or the contrast** of these might be useful (e.g. the second example has a lot of colors, and the fourth one has too little saturation, so I increased the saturation and the contrast and decreased overall brightness to be able to sample good colors.)

Giving the user a choice of a "target" contrast and saturation could also be a good idea.

We need to pick a color with low lightness and low saturation for the background color, and a color with medium-high lightness and high saturation for the foreground color, then use those for the sparse and dense areas of the graph respectively. Here are the above examples with the actual bar, with varying levels of _Background dim_ and _blur_:

![modified difficulty bar for "Tower of Heaven", "Extra" difficulty with 0% Background dim and no Background blur](https://github.com/ppy/osu/assets/72462464/eec2d06f-81be-4a0d-9919-1851b8eef510)
![modified difficulty bar for "Tower of Heaven", "Extra" difficulty with 0% Background dim and 100% Background blur](https://github.com/ppy/osu/assets/72462464/89583383-630f-48f0-a1c8-e4f329586281)
![modified difficulty bar for "Tower of Heaven", "Extra" difficulty with 30% Background dim and no Background blur](https://github.com/ppy/osu/assets/72462464/03b84ce4-93f8-4aec-b974-1adc251570ed)
![modified difficulty bar for "Tower of Heaven", "Extra" difficulty with 30% Background dim and 100% Background blur](https://github.com/ppy/osu/assets/72462464/155192e1-798b-422d-a57e-511f77827006)
![modified difficulty bar for "Tower of Heaven", "Extra" difficulty with default (70%) Background dim and no Background blur](https://github.com/ppy/osu/assets/72462464/c9a4313d-6d72-43e4-91d4-15097fc5756c)
![modified difficulty bar for "Tower of Heaven", "Extra" difficulty with default (70%) Background dim and 100% Background blur](https://github.com/ppy/osu/assets/72462464/00ef6e2e-8241-4cba-817b-02c0005aec64)
![modified difficulty bar for "Tower of Heaven", "Extra" difficulty with 100% Background dim](https://github.com/ppy/osu/assets/72462464/319efaf2-ba5d-4c59-be9a-8607d123a24f)

![modified difficulty bar for "Airman ga Taosenai", "Skystar's Extra" difficulty with 0% Background dim and no Background blur](https://github.com/ppy/osu/assets/72462464/6e55ad45-391d-4528-87d9-9dcfa54eed02)
![modified difficulty bar for "Airman ga Taosenai", "Skystar's Extra" difficulty with 0% Background dim and 100% Background blur](https://github.com/ppy/osu/assets/72462464/8360d98b-08cc-4edf-83e6-479539675129)
![modified difficulty bar for "Airman ga Taosenai", "Skystar's Extra" difficulty with 30% Background dim and no Background blur](https://github.com/ppy/osu/assets/72462464/d6cc92cc-c725-4fa7-817c-c4716e3a4a87)
![modified difficulty bar for "Airman ga Taosenai", "Skystar's Extra" difficulty with 30% Background dim and 100% Background blur](https://github.com/ppy/osu/assets/72462464/2d83b45d-95b7-48cd-ba9c-a9677fef3d28)
![modified difficulty bar for "Airman ga Taosenai", "Skystar's Extra" difficulty with default (70%) Background dim and no Background blur](https://github.com/ppy/osu/assets/72462464/328184c0-af98-44e0-9457-a70b41b733e1)
![modified difficulty bar for "Airman ga Taosenai", "Skystar's Extra" difficulty with default (70%) Background dim and 100% Background blur](https://github.com/ppy/osu/assets/72462464/9819c198-1fe4-45f1-9593-a531f0432566)
![modified difficulty bar for "Airman ga Taosenai", "Skystar's Extra" difficulty with 100% Background dim](https://github.com/ppy/osu/assets/72462464/8eadd814-ac6b-4427-8d4e-80cd5a0a735d)

![modified difficulty bar for "ILY" (by Panda Eyes), "Fanteer's Extreme" difficulty with 0% Background dim and no Background blur](https://github.com/ppy/osu/assets/72462464/3ab6c6d9-27c2-4476-8e1f-74d214ccca43)
![modified difficulty bar for "ILY" (by Panda Eyes), "Fanteer's Extreme" difficulty with 0% Background dim and 100% Background blur](https://github.com/ppy/osu/assets/72462464/eb07c9e6-4f93-4fcc-9b08-ca6450a65dfc)
![modified difficulty bar for "ILY" (by Panda Eyes), "Fanteer's Extreme" difficulty with 30% Background dim and no Background blur](https://github.com/ppy/osu/assets/72462464/378adb74-4302-43eb-bff3-b328aa8232df)
![modified difficulty bar for "ILY" (by Panda Eyes), "Fanteer's Extreme" difficulty with 30% Background dim and 100% Background blur](https://github.com/ppy/osu/assets/72462464/42edfecf-2493-4bfa-991c-5697129db2d9)
![modified difficulty bar for "ILY" (by Panda Eyes), "Fanteer's Extreme" difficulty with default (70%) Background dim and no Background blur](https://github.com/ppy/osu/assets/72462464/fa82295a-3735-480f-8579-682d57de6aad)
![modified difficulty bar for "ILY" (by Panda Eyes), "Fanteer's Extreme" difficulty with default (70%) Background dim and 100% Background blur](https://github.com/ppy/osu/assets/72462464/9046d481-dd26-4ede-9f7d-4012518d287a)
![modified difficulty bar for "ILY" (by Panda Eyes), "Fanteer's Extreme" difficulty with 100% Background dim](https://github.com/ppy/osu/assets/72462464/3e97bfe5-d2ef-4d3f-bb7c-b84dae62c819)

![modified difficulty bar for "Ooi" (by O2i3), "Extreme" difficulty with 0% Background dim and no Background blur](https://github.com/ppy/osu/assets/72462464/d13ec6fe-7bb2-4e7a-a3dd-84e285cbd97a)
![modified difficulty bar for "Ooi" (by O2i3), "Extreme" difficulty with 0% Background dim and 100% Background blur](https://github.com/ppy/osu/assets/72462464/1ddcb545-428f-40d6-bcc2-3cf98ac95f28)
![modified difficulty bar for "Ooi" (by O2i3), "Extreme" difficulty with 30% Background dim and no Background blur](https://github.com/ppy/osu/assets/72462464/7916d61d-8dfa-48c2-ac94-b28edb57bc82)
![modified difficulty bar for "Ooi" (by O2i3), "Extreme" difficulty with 30% Background dim and 100% Background blur](https://github.com/ppy/osu/assets/72462464/69389f1d-ceab-48af-85ec-f3ac16343c82)
![modified difficulty bar for "Ooi" (by O2i3), "Extreme" difficulty with default (70%) Background dim and no Background blur](https://github.com/ppy/osu/assets/72462464/f5e1882e-cb0e-4399-95a8-fa0ca2de86e4)
![modified difficulty bar for "Ooi" (by O2i3), "Extreme" difficulty with default (70%) Background dim and 100% Background blur](https://github.com/ppy/osu/assets/72462464/a8ca376b-0381-4eba-8285-7974a5986cc3)
![modified difficulty bar for "Ooi" (by O2i3), "Extreme" difficulty with 100% Background dim](https://github.com/ppy/osu/assets/72462464/b8d8c580-0bbc-4157-8882-7dc20f93475d)

This does need additional **programmatical implementation** (and I admit I just used the color picker manually), but it generally results in better contrast against most backgrounds.

As much as you love **transparent** UI elements, I think it **should be avoided** here, or at best, applied to the _background color_ (the darker one) _and in moderation_.

To "idiot-proof" this and accomodate for backgrounds with _a lot_ of colors, an outline or drop shadow could be used.

For storyboards or videos, _I don't know_. We could hope that the coloring of those matches the static background (_we shouldn't pull a YouTube Ambient mode here_).

### Draw an actual area graph

As in, increase the height of the bar according to the density.

This might be very tedious to implement, but it would provide the best accessibility.

_(And because it is tedious, I'll only try to make an example image if this is even plausible.)_

## Other considerations

The bar obviously should be colored differently for the already completed parts (ie. before the current position). You could achieve this by pushing color values towards brighter colors, shifting the hue (in case it is colored) or as in #5681.

Unless these suggestions fit current or future designs, if implemented, they should be completely optional or presented as accessibility options (in the skin editor).

**Comment by fyyrraaa** [comment]
AFAIK previously it wasn't this this low-contrast upon release ([2023.121.0](https://github.com/ppy/osu/releases/tag/2023.121.0), see #22144); but since [2023.621.0](https://github.com/ppy/osu/releases/tag/2023.621.0) the progress bar was changed so that it would be more visible in bright backgrounds (see #23884), as far as I understand. It's something to keep in mind in this case; but in any case I do agree with the fact that the density graph in Argon is not really readable.

**Reply by peppy** [reply]
Yep, when proposing changes, please also test against background dim at 0% or 30%.

**Reply by verarr** [reply]
Apologies for the oversight, I've added a new section at the end of "_Use color instead of opacity_" detailing some more ideas.

TLDR.: _Black and White_ works pretty much anywhere, except for pure white and black backgrounds. To use colors, we could pick colors dynamically based on a color inverted version of the background. I've included some examples.

**Reply by verarr** [reply]
Just in case there's any ambiguity about this: I am not a graphic designer at all, and the images I made for this probably suck as they are right now. They are just ideas on where you might want to look in terms of the readability of the progress bar.

**Reply by peppy** [reply]
You can't just make the background of the bar black. It's completely changing the visual style.

**Reply by fyyrraaa** [reply]
I think for the benefit of the discussion thread I'll link the current Argon HUD design to serve as reference: https://www.figma.com/file/Tu1C6CR2XOdhm7kxq1JQNj/Gameplay-Update (other people might've not seen this yet)

**Reply by verarr** [reply]
I know. Maybe someone else is able to salvage them and turn them into something you'd be happy with.

Let's maybe include a message at the top of this post about this?

Though, I think it's agreed that it has some issues with readability.

**Reply by peppy** [reply]
Yep, it's definitely not in a perfect state yet from a readability perspective.

I do like the idea of pulling colours from the beatmap background. It might be something we can do across the argon theme, in fact.

Just need to find a way to add some more transparency back.
